hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/ha/HASecurityUtils.java

```diff
@@ -0,0 +1,291 @@
+    CertificateClient.InitResponse response = certClient.init();
+    LOG.info("Init response: {}", response);
+    switch (response) {
+    case SUCCESS:
+      LOG.info("Initialization successful.");
+      break;
+    case GETCERT:
+      if (!primaryscm) {
+        getSCMSignedCert(certClient, conf, fetchedScmId, scmStorageConfig,
+            scmAddress);
+      } else {
+        getPrimarySCMSignedCert(certClient, conf, fetchedScmId,
+            scmStorageConfig, scmAddress);
+      }
+      LOG.info("Successfully stored SCM signed certificate.");
+      break;
+    case FAILURE:
+      LOG.error("SCM security initialization failed.");
+      throw new RuntimeException("OM security initialization failed.");
+    case RECOVER:
+      LOG.error("SCM security initialization failed. SCM certificate is " +
+          "missing.");
+      throw new RuntimeException("SCM security initialization failed.");
+    default:
+      LOG.error("SCM security initialization failed. Init response: {}",
+          response);
+      throw new RuntimeException("SCM security initialization failed.");
+    }
+  }
+
+  /**
+   * Get SCM signed certificate and store it using certificate client.
+   */
+  private static void getSCMSignedCert(CertificateClient client,
+      OzoneConfiguration config, String fetchedSCMId,
+      SCMStorageConfig scmStorageConfig, InetSocketAddress scmAddress)
+      throws IOException {
+
+    PKCS10CertificationRequest csr = generateCSR(client, scmStorageConfig,
+        config, scmAddress, fetchedSCMId);
+
+    HddsProtos.ScmNodeDetailsProto scmNodeDetailsProto =
+        HddsProtos.ScmNodeDetailsProto.newBuilder()
+            .setClusterId(scmStorageConfig.getClusterID())
+            .setHostName(scmAddress.getHostName())
+            .setScmNodeId(fetchedSCMId).build();
+
+    SCMSecurityProtocolClientSideTranslatorPB secureScmClient =
+        HddsServerUtil.getScmSecurityClient(config);
+
+    SCMGetCertResponseProto response = secureScmClient.
+        getSCMCertChain(scmNodeDetailsProto, getEncodedString(csr));
+    String pemEncodedCert = response.getX509Certificate();
+
+    try {
+
+      // Store SCM CA certificate.
+      if (response.hasX509CACertificate()) {
+        String pemEncodedRootCert = response.getX509CACertificate();
+        client.storeCertificate(pemEncodedRootCert, true, true);
+        client.storeCertificate(pemEncodedCert, true);
+
+
+        // Write to the location, so that Default CA server can read the
+        // certificate generated by cert client.
+        CertificateCodec certCodec =
+            new CertificateCodec(new SecurityConfig(config),
+                client.getComponentName());
+
+        X509Certificate certificate =
+            CertificateCodec.getX509Certificate(pemEncodedCert);
+        certCodec.writeCertificate(certCodec.getCertificateHolder(certificate));
+
+        // Persist scm cert serial ID.
+        scmStorageConfig.setScmCertSerialId(certificate.getSerialNumber()
```

Comment:
On primary SCM(Root CA), the non-root SCM cert will overwrite the cert serial id in the normal run. If it is expected, should we skip write the root CA cert serial id to the version file?

Reply:
On all SCM's only persisting sub SCM cert SerialID to version file. This is to find out after a restart, what is the Certificate SerialID and so that getCertificate() will work. In all SCMS (primary/bootstrap) followed same approach

So in SCMCertClient

certSerialID -> Certificate got from root CA
caCertId -> Root CA Certificate 

Not persisted rootCA cert again, as for SCM caCertId and rootCaCertID are same.

Whereas for other roles like OM/DN

certSerialID -> Certificate got from sub SCM CA(any SCM which is leader at that point)
caCertId -> Sub SCM CA Certificate ID
rootCaCertId -> root SCM CA Certificate ID